File: frontend/src/lib/api/de-approval.ts
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";

import { api } from "./client";
import type { ProjectStatus } from "./projects";

// DE Project Approval (design-reference/de-approval) — the allocated Delivery
// Excellence assessor reviews governance completeness module by module and
// Approves or Returns the project. The queue is scoped server-side to the
// signed-in DE's allocations; period_id is a display filter only.

export type DeReviewStatus = "In Review" | "Returned" | "Approved";
export type DeModuleReviewAction = "Not Reviewed" | "Reviewed" | "Gap Identified";
export type GovernanceModuleKey =
  | "project_profile"
  | "scope_schedule"
  | "map_oracle_projects"
  | "contractual_compliance"
  | "raido"
  | "measurement";

export const MODULE_REVIEW_ACTIONS: DeModuleReviewAction[] = [
  "Not Reviewed",
  "Reviewed",
  "Gap Identified",
];

// Where each checklist row's "View" link goes — dedicated read-only
// /de-approval/{projectId}/{path} module views owned by this feature (no edit
// affordances, no editable sibling route).
export const MODULE_VIEW_PATH: Record<GovernanceModuleKey, string> = {
  project_profile: "project-profile",
  scope_schedule: "scope-schedule",
  map_oracle_projects: "oracle-mapping",
  contractual_compliance: "contractual",
  raido: "raido",
  measurement: "measurement",
};

export type GovernanceModuleStatus = {
  key: GovernanceModuleKey;
  label: string;
  mandatory: boolean;
  complete: boolean;
  gaps: string | null;
  review_action: DeModuleReviewAction;
  last_updated: string | null;
  // Partial progress: required fields filled / required fields expected.
  fields_complete: number;
  fields_total: number;
  progress_pct: number;
};

export type GovernanceCompleteness = {
  completion_pct: number;
  modules_complete: number;
  modules_incomplete: number;
  gaps_count: number;
  critical_gaps: number;
  modules: GovernanceModuleStatus[];
};

export type DeApprovalKpis = {
  awaiting_review: number;
  in_review: number;
  returned: number;
};

export type DeApprovalQueueRow = {
  project_id: string;
  project_code: string;
  project_name: string;
  account_name: string | null;
  geo_name: string | null;
  region_name: string | null;
  project_type_name: string | null;
  project_manager_name: string | null;
  completion_pct: number;
  gaps_count: number;
  project_status: ProjectStatus;
  de_review_status: DeReviewStatus | null;
  last_updated: string;
  href: string;
};

export type DeApprovalQueueResponse = {
  period_id: string | null;
  kpis: DeApprovalKpis;
  rows: DeApprovalQueueRow[];
};

export type DeReviewDetail = {
  project_id: string;
  project_code: string;
  project_name: string;
  account_name: string | null;
  project_manager_name: string | null;
  project_status: ProjectStatus;
  de_review_status: DeReviewStatus | null;
  de_review_remarks: string | null;
  de_reviewed_by: string | null;
  de_reviewed_at: string | null;
  completeness: GovernanceCompleteness;
};

export type DeReviewDecision = "Approve" | "Return";

export function useDeApprovalQueue(periodId: string | null) {
  return useQuery({
    queryKey: ["de-approval-queue", periodId],
    queryFn: () =>
      api.get<DeApprovalQueueResponse>(
        periodId ? `/de-approval/queue?period_id=${periodId}` : "/de-approval/queue",
      ),
  });
}

export function useDeReviewDetail(projectId: string | null) {
  return useQuery({
    queryKey: ["de-approval-detail", projectId],
    queryFn: () => api.get<DeReviewDetail>(`/de-approval/${projectId}`),
    enabled: !!projectId,
  });
}

export function useUpdateModuleReview(projectId: string | null) {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: ({
      moduleKey,
      reviewAction,
    }: {
      moduleKey: GovernanceModuleKey;
      reviewAction: DeModuleReviewAction;
    }) =>
      api.put<GovernanceModuleStatus>(`/de-approval/${projectId}/modules/${moduleKey}`, {
        review_action: reviewAction,
      }),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ["de-approval-detail", projectId] });
      queryClient.invalidateQueries({ queryKey: ["de-approval-queue"] });
    },
  });
}

export function useDeReviewDecision(projectId: string | null) {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (payload: { decision: DeReviewDecision; remarks: string; reviewed_by: string }) =>
      api.patch<DeReviewDetail>(`/de-approval/${projectId}/decision`, payload),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ["de-approval-detail", projectId] });
      queryClient.invalidateQueries({ queryKey: ["de-approval-queue"] });
      queryClient.invalidateQueries({ queryKey: ["project", projectId] });
      queryClient.invalidateQueries({ queryKey: ["projects"] });
      queryClient.invalidateQueries({ queryKey: ["dashboard-de-summary"] });
    },
  });
}

File: backend/app/schemas/de_approval.py
"""Schemas for DE Project Allocation + DE Project Approval
(design-reference/de-approval). Allocation assigns Project.delivery_excellence_id;
approval runs a project_status Pending Approval -> Approved / Draft transition
plus a de_review_status sub-state and a per-module review checklist."""

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field

from app.schemas.enums import (
    DeModuleReviewAction,
    DeReviewStatus,
    GovernanceModuleKey,
    ProjectLifecycleStatus,
    ProjectStatus,
)

# --- Governance completeness -------------------------------------------------


class GovernanceModuleStatus(BaseModel):
    key: GovernanceModuleKey
    label: str
    mandatory: bool
    complete: bool
    gaps: str | None = None
    review_action: DeModuleReviewAction = DeModuleReviewAction.NOT_REVIEWED
    last_updated: datetime | None = None
    # Partial progress: required fields filled / required fields expected.
    fields_complete: int = 0
    fields_total: int = 0
    progress_pct: int = 0


class GovernanceCompleteness(BaseModel):
    completion_pct: int
    modules_complete: int
    modules_incomplete: int
    gaps_count: int
    critical_gaps: int
    modules: list[GovernanceModuleStatus]


# --- DE Project Allocation -------------------------------------------------


class DeAllocationRow(BaseModel):
    project_id: UUID
    project_code: str
    project_name: str
    account_name: str | None = None
    project_manager_name: str | None = None
    project_status: ProjectStatus
    lifecycle_status: ProjectLifecycleStatus | None = None
    delivery_excellence_id: UUID | None = None
    delivery_excellence_name: str | None = None
    de_allocated_at: datetime | None = None
    completion_pct: int
    gaps_count: int


class DeAllocationAssignment(BaseModel):
    project_id: UUID
    delivery_excellence_id: UUID


class DeAllocationBulkAssign(BaseModel):
    assignments: list[DeAllocationAssignment] = Field(min_length=1)


# --- DE Project Approval queue -------------------------------------------------


class DeApprovalKpis(BaseModel):
    awaiting_review: int
    in_review: int
    returned: int


class DeApprovalQueueRow(BaseModel):
    project_id: UUID
    project_code: str
    project_name: str
    account_name: str | None = None
    geo_name: str | None = None
    region_name: str | None = None
    project_type_name: str | None = None
    project_manager_name: str | None = None
    completion_pct: int
    gaps_count: int
    project_status: ProjectStatus
    lifecycle_status: ProjectLifecycleStatus | None = None
    de_review_status: DeReviewStatus | None = None
    last_updated: datetime
    href: str


class DeApprovalQueueResponse(BaseModel):
    period_id: UUID | None = None
    kpis: DeApprovalKpis
    rows: list[DeApprovalQueueRow]


# --- Project Governance Review workspace -------------------------------------------------


class DeReviewDetail(BaseModel):
    project_id: UUID
    project_code: str
    project_name: str
    account_name: str | None = None
    project_manager_name: str | None = None
    project_status: ProjectStatus
    lifecycle_status: ProjectLifecycleStatus | None = None
    de_review_status: DeReviewStatus | None = None
    de_review_remarks: str | None = None
    de_reviewed_by: UUID | None = None
    de_reviewed_at: datetime | None = None
    completeness: GovernanceCompleteness


class DeModuleReviewUpdate(BaseModel):
    review_action: DeModuleReviewAction
    remarks: str | None = None


class DeReviewDecisionRequest(BaseModel):
    decision: Literal["Approve", "Return"]
    remarks: str = Field(min_length=1)  # DE Review Remarks — mandatory
    reviewed_by: UUID
